pkg/user/registry/useridentitymapping/rest.go

```diff
@@ -4,42 +4,360 @@ import (
+	if len(mapping.ResourceVersion) > 0 && mapping.ResourceVersion != oldMapping.ResourceVersion {
+		return nil, false, kerrs.NewConflict("UserIdentityMapping", mapping.Name, fmt.Errorf("the resource was updated to %s", oldMapping.ResourceVersion))
+	}
+
+	// If we're "updating" to the user we're already pointing to, we're already done
+	if mapping.User.Name == oldUser.Name {
+		return oldMapping, false, nil
+	}
+
+	// Get new user
+	newUser, err := s.userRegistry.GetUser(ctx, mapping.User.Name)
+	if kerrs.IsNotFound(err) {
+		errs := kerrs.ValidationErrorList([]error{
+			kerrs.NewFieldInvalid("user.name", mapping.User.Name, "referenced user does not exist"),
+		})
+		return nil, false, kerrs.NewInvalid("UserIdentityMapping", mapping.Name, errs)
+	}
+	if err != nil {
+		return nil, false, err
+	}
+
+	// Update the new user to point at the identity. If this fails, Update is re-entrant
+	if addIdentityToUser(identity, newUser) {
+		if _, err := s.userRegistry.UpdateUser(ctx, newUser); err != nil {
+			return nil, false, err
+		}
+	}
+
+	// Update the identity to point at the new user. If this fails. Update is re-entrant
+	if setIdentityUser(identity, newUser) {
+		if updatedIdentity, err := s.identityRegistry.UpdateIdentity(ctx, identity); err != nil {
+			return nil, false, err
+		} else {
+			identity = updatedIdentity
+		}
+	}
+
+	// At this point, the mapping for the identity has been updated to the new user
+	// Everything past this point is cleanup
+
+	// Update the old user to no longer point at the identity.
+	// If this fails, log the error, but continue, because Update is no longer re-entrant
+	if removeIdentityFromUser(identity, oldUser) {
+		if _, err := s.userRegistry.UpdateUser(ctx, oldUser); err != nil {
+			glog.Errorf("Error removing identity reference %s from user %s: %v", identity.Name, oldUser.Name, err)
+		}
+	}
+
+	return mappingFor(newUser, identity), false, nil
+}
+
+func (s *REST) Delete(ctx kapi.Context, name string) (runtime.Object, error) {
+	identity, _, user, _, _, mappingErr := s.getRelatedObjects(ctx, name)
+
+	if mappingErr != nil {
+		return nil, mappingErr
+	}
+
+	// Disassociate the identity with the user first
+	// If this fails, Delete is re-entrant
+	if removeIdentityFromUser(identity, user) {
+		if _, err := s.userRegistry.UpdateUser(ctx, user); err != nil {
+			return nil, err
+		}
+	}
+
+	// Remove the user association from the identity last.
+	// If this fails, log the error, but continue, because Delete is no longer re-entrant
+	// At this point, the mapping for the identity no longer exists
+	if unsetIdentityUser(identity) {
+		if _, err := s.identityRegistry.UpdateIdentity(ctx, identity); err != nil {
+			glog.Errorf("Error removing user reference %s from identity %s: %v", user.Name, identity.Name, err)
+		}
+	}
+
+	return &kapi.Status{Status: kapi.StatusSuccess}, nil
+}
+
+func (s *REST) getRelatedObjects(ctx kapi.Context, name string) (
```

Comment:
Why? This requires user and identity to already exist. Provision is read-only if identity already exists. I agree both files are messy, but they shouldn't interact weirdly.